src/main.cpp

```diff
@@ -4619,54 +4600,72 @@ bool SendMessages(CNode* pto, bool fSendTrickle)
+        
+        if (pto->nNextInvSend < nNowMicros)
+            vInvWait.reserve(pto->vInventoryToSend.size());
 
-                    if (fTrickleWait)
+        {
+            LOCK(pto->cs_inventory);
+            
+            // TODO it would be sufficient to walk vInventoryToSend randomly, shuffling is relatively inefficient
+            std::random_shuffle(pto->vInventoryToSend.begin(), pto->vInventoryToSend.end());
+
+            BOOST_FOREACH(const CInv& inv, pto->vInventoryToSend) {
+                switch(inv.type) {
+                    case MSG_BLOCK:
+                    case MSG_FILTERED_BLOCK:
                     {
-                        vInvWait.push_back(inv);
-                        continue;
+                        // Aggressively push MSG_BLOCK/MSG_FILTERED_BLOCK
+                        if (pto->setInventoryKnown.insert(inv).second) {
+                            vInv.push_back(inv);
+                            pto->PushMessage("inv", vInv);
+                            vInv.clear();
+                        }
+                        break;
                     }
-                }
+                    case MSG_TX:
+                    {
+                        if (pto->nNextInvSend < nNowMicros || pto->fWhitelisted) {
+                            if (pto->setInventoryKnown.insert(inv).second) {
+                                vInv.push_back(inv);
+                                if (vInv.size() >= 1000) {
+                                    pto->PushMessage("inv", vInv);
+                                    vInv.clear();
+                                }
+                            }
+                        }
+                        else
+                            vInvWait.push_back(inv);
 
-                // returns true if wasn't already contained in the set
-                if (pto->setInventoryKnown.insert(inv).second)
-                {
-                    vInv.push_back(inv);
-                    if (vInv.size() >= 1000)
+                        break;
+                    }
+                    default:
                     {
-                        pto->PushMessage("inv", vInv);
-                        vInv.clear();
+                        vInv.push_back(inv);
+                        if (vInv.size() >= 1000) {
+                            pto->PushMessage("inv", vInv);
+                            vInv.clear();
+                        }
                     }
                 }
             }
+
             pto->vInventoryToSend = vInvWait;
         }
+
         if (!vInv.empty())
             pto->PushMessage("inv", vInv);
 
+        if (pto->nNextInvSend < nNowMicros)
+            pto->nNextInvSend = nNowMicros + GetRand(10 * 1000000);
+
         // Detect whether we're stalling
-        int64_t nNow = GetTimeMicros();
-        if (!pto->fDisconnect && state.nStallingSince && state.nStallingSince < nNow - 1000000 * BLOCK_STALLING_TIMEOUT) {
+        nNowMicros = GetTimeMicros();
```

Comment:
If the difference in time between them is more than a few milliseconds, we have different problems :)